### ISSUE TITLE
Prepare for upstream changes. [phab:D625]

### ISSUE THREAD

**Issue by kythe** [google/kythe#886]
_Created by **zarko** at 2015-12-02 15:24:40_: 

Current clang's RecursiveASTVisitor API has changed. To make it easier to update in the future, we'll update our code now (especially given that doing so doesn't break any tests).

Without this change, we fail to compile:
```
kythe/cxx/indexer/cxx/IndexerASTHooks.h:221:12: error: no matching member function for call to 'TraverseNode'
    return TraverseNode(StmtNode, &VisitorBase::TraverseStmt);
           ^~~~~~~~~~~~
kythe/cxx/indexer/cxx/IndexerASTHooks.h:168:8: note: candidate template ignored: failed template argument deduction
  bool TraverseNode(T *Node, bool (VisitorBase::*traverse)(T *)) {
       ^
1 error generated.
```

This happened due to Clang commit 89d5b46c217bb7fd5d23e92def8f9a6edd57090a. The change mirrors the change that commit makes to ParentMapASTVisitor.

**Comment by kythe** [comment]
_**fromberger** wrote at 2015-12-02 16:12:10_:

Can you please include in the change description a bit more about what is changing, and why this fixes it?
Doesn't need to be a novel, but a sentence or two would be appreciated.

**Comment by kythe** [comment]
_**fromberger** commented inline (context lost during migration) at 2015-12-02 16:22:25_:

Does the Clang change really require that we template on the function type here? Maybe I should be looking at whatever upstream changes you're presponding to here, but I'm leery of templating against an arbitrary function type. At the very least I'd like to see some documentation of that template parameter.

**Comment by kythe** [comment]
_**schroederc** accepted this revision at 2015-12-02 19:34:33_

**Comment by kythe** [comment]
_**zarko** wrote at 2015-12-10 18:10:02_:

  - Comment type arguments.

**Comment by kythe** [comment]
_**zarko** commented inline (context lost during migration) at 2015-12-10 18:10:38_:

PTAL

**Comment by kythe** [comment]
_**schroederc** wrote at 2015-12-10 18:11:23_:

LGTM

**Comment by kythe** [comment]
_**fromberger** accepted this revision at 2015-12-10 18:12:13_

**Comment by kythe** [comment]
_**zarko** closed via commit 3e00cb3be4d634d3162652bd393e3ad71b4db350 at 2015-12-10 18:12:45_
